src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateAtTheMoment = { ...state }; // clone the state at the beginning
+  const stateHistory = []; // const, because an array is an object in JS
+
+  for (const action of actions) {
+    let newState = { ...stateAtTheMoment };
+    // Clone the current state to the new state
+
+    if (action.type === 'clear') {
+      newState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
+
+    stateHistory.push(newState);
```

Comment:
You are currently pushing the `newState` into the `stateHistory` array within the loop. This action is repeated in all cases and should be done once after the switch statement to adhere to the DRY (Don't Repeat Yourself) principle.